src/providers/MeetingProvider/MeetingManager.ts

```diff
@@ -286,9 +290,19 @@ export class MeetingManager implements AudioVideoObserver {
     let callback: DeviceLabelTrigger;
 
     if (typeof deviceLabels === 'function') {
-      callback = deviceLabels;
+      callback = async (): Promise<MediaStream> => {
+        const stream = await deviceLabels();
+        this.hasAudioPermission = !!stream.getAudioTracks().length;
+        this.hasVideoPermission = !!stream.getVideoTracks().length;
+        return stream;
+      };
     } else if (deviceLabels === DeviceLabels.None) {
-      callback = noOpDeviceLabelHook;
+      callback = async (): Promise<MediaStream> => {
+        const stream = await noOpDeviceLabelHook();
+        this.hasAudioPermission = !!stream.getAudioTracks().length;
+        this.hasVideoPermission = !!stream.getVideoTracks().length;
+        return stream;
+      };
```

Comment:
By definition the new fields will be false here — an empty stream has no tracks.

What's supposed to happen if the builder uses a hook (say, camera-only), and then switches to `DeviceLabels.None`? Do we keep `hasVideoPermission = true`, or do we set it to false? That applies more generally to the case when they use any non-subset value, such as audio and then video.

Reply:
#### Case 1

1. call `mm.join(DeviceLabels.Audio)`, and redirect to device page
    * `deviceLabelTrigger` is triggered, browser ask permission for **microphone**
    *  Allow the permission
    * `hasAudioPermission = true`, `hasVideoPermission = false`
2. redirect to meeting page
3. call `mm.invokeDeviceProvider(DeviceLabels.Video)`
    * the `mm.setupDeviceLabelTrigger(DeviceLabels.Video)` is called
        * then `deviceLabelTrigger` is triggered, browser ask permission for **camera**
            * `hasAudioPermission = false`, `hasVideoPermission = true`
            * `hasAudioPermission` is set to false, cause this time returned `stream` has no `AudioTrack`

#### Case 2

1. call `mm.join(DeviceLabels.Audio)`, and redirect to device page
    * `deviceLabelTrigger` is triggered, browser ask permission for **microphone**
        *  Allow the permission
        * `hasAudioPermission = true`, `hasVideoPermission = false`
2. redirect to meeting page
3. call `mm.invokeDeviceProvider(DeviceLabels.None)`
    * the `mm.setupDeviceLabelTrigger(DeviceLabels.None)` is called
        * (Chrome) the `noOpDeviceLabelHook` `deviceLabelTrigger` is triggered, no permission prompt
            * `hasAudioPermission = false`, `hasVideoPermission = false`
            * update the value, cause returned `stream` has no tracks
        * (Firefox/Safari) `deviceLabelTrigger` is **NOT** triggered
            * `hasAudioPermission = true`, `hasVideoPermission = false`
            * keep the previous value

Reply:
I would say the essence of these two variables is a helper variable **ONLY** used to check whether or not to select one kind of devices in `listAndSelectDevices()`. But to share these variables accross `setupDeviceLabelTrigger` and `listAndSelectDevices`, we have to add them as a property of `MeetingManager`, thus we need to make their value meaningful. 

It feels like a bad design to me. I think we have following solutions:
* Should we add comments to explain the responsibility of these two variables to address this issue?
* Should we do some extra work to keep their value meaningful? I don't think it's good to keep the browser's property in `MeetingManager` as you mentioned. Especially the behavior of browser varies.
* Avoid adding them as a property, and find another way to do it.